NewEssentials/Players/AfkChecker.cs

```diff
@@ -0,0 +1,110 @@
+        private readonly IConfiguration m_Config;
+        private readonly NewEssentials m_Plugin;
+        private readonly IUserManager m_UserManager;
+        
+        public AfkChecker(IEventBus bus, IConfiguration config, IPluginAccessor<NewEssentials> plugin, IUserManager users)
+        {
+            if (!config.GetValue<bool>("afkchecker:enabled"))
+                return;
+            m_Users = new Dictionary<IUser, TimeSpan>();
+            
+            bus.Subscribe<UserConnectedEvent>(plugin.Instance, (provider, sender, @event) => PlayerJoin(@event));
+            bus.Subscribe<UserConnectedEvent>(plugin.Instance, (provider, sender, @event) => PlayerLeave(@event));
+            
+            
+            AsyncHelper.Schedule("NewEssentials::AfkChecker", async () => await CheckAfkPlayers());
+            m_Events = bus;
+            m_Config = config;
+            m_Plugin = plugin.Instance;
+            m_UserManager = users;
+        }
+
+        private async Task PlayerJoin(UserConnectedEvent @event)
+        {
+            if (m_Users.ContainsKey((UnturnedUser)@event.User))
+            {
+                m_Users[(UnturnedUser)@event.User] = DateTime.Now.TimeOfDay;
+                return;
+            }
+            
+            m_Users.Add((UnturnedUser)@event.User, DateTime.Now.TimeOfDay);
+            PlayerMovementCheckerComponent component = ((UnturnedUser) @event.User).Player.gameObject.AddComponent<PlayerMovementCheckerComponent>();
+            component.Resolve(this);
+
+        }
+
+        private async UniTask CheckAfkPlayers()
+        {
+            while (m_Plugin.IsComponentAlive)
+            {
+                UniTask.Delay(3000);
+
+                IUser user;
+                
+                // ReSharper disable once ForCanBeConvertedToForeach
+                for (int i = 0; i < Provider.clients.Count; i++)
+                {
+                    user = await m_UserManager.FindUserAsync(KnownActorTypes.Player,
+                        Provider.clients[i].playerID.playerName, UserSearchMode.Name);
+                    if (DateTime.Now.TimeOfDay.TotalSeconds - m_Users[user].TotalSeconds >=
+                        m_Config.GetValue<float>("afkchecker:totalseconds"))
+                        await ((UnturnedUser) user).KickAsync("You were kicked for being afk!");
+                }
+            }
+        }
+        
+        private async Task PlayerLeave(UserConnectedEvent @event)
+        {
+            if (m_Users.ContainsKey((UnturnedUser)@event.User))
+                m_Users.Remove((UnturnedUser)@event.User);
+        }
+
+
+        public async UniTask UpdateUser(IUser user)
+        {
+            if (m_Users.ContainsKey(user))
+                m_Users[user] = DateTime.Now.TimeOfDay;
+        }
+
+        public void UpdatePlayer(Player player)
+        {
+            IUser user = m_UserManager.FindUserAsync(KnownActorTypes.Player, player.channel.owner.playerID.playerName,
```

Comment:
Must be async with await, also find user by ID not name

Reply:
Async voids shouldn't be used, what would you recommmend?

Reply:
async Task

Reply:
Nevermind, saw your later review